File: clients/core-ts/src/pow.test.ts
import { describe, expect, it } from 'vitest';
import { hasLeadingZeroBits, powHash, solvePoW, verifyPoW } from './pow.js';

describe('hasLeadingZeroBits', () => {
  it('checks whole leading zero bytes', () => {
    expect(hasLeadingZeroBits(new Uint8Array([0, 0, 1]), 16)).toBe(true);
    expect(hasLeadingZeroBits(new Uint8Array([0, 1, 0]), 16)).toBe(false);
  });

  it('checks a partial byte with the correct mask', () => {
    // 20 bits => 2 full zero bytes + top nibble of byte 3 zero.
    expect(hasLeadingZeroBits(new Uint8Array([0, 0, 0x0f]), 20)).toBe(true);
    expect(hasLeadingZeroBits(new Uint8Array([0, 0, 0x10]), 20)).toBe(false);
  });

  it('treats 0 bits as always satisfied', () => {
    expect(hasLeadingZeroBits(new Uint8Array([0xff]), 0)).toBe(true);
  });
});

describe('powHash', () => {
  it('is deterministic and concatenates utf8(pub)||utf8(nonce)', () => {
    const a = powHash('pubkey', '42');
    const b = powHash('pubkey', '42');
    expect(a).toEqual(b);
    expect(a).toHaveLength(32);
    expect(powHash('pubkey', '43')).not.toEqual(a);
  });
});

describe('solvePoW / verifyPoW', () => {
  const pub = 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=';

  it('finds a nonce that satisfies the difficulty', () => {
    const nonce = solvePoW(pub, 12);
    expect(nonce).toMatch(/^\d+$/);
    expect(hasLeadingZeroBits(powHash(pub, nonce), 12)).toBe(true);
    expect(verifyPoW(pub, nonce, 12)).toBe(true);
  });

  it('with an explicit start of 0, finds the smallest qualifying nonce', () => {
    const nonce = solvePoW(pub, 10, 0);
    const n = Number(nonce);
    for (let i = 0; i < n; i++) {
      expect(hasLeadingZeroBits(powHash(pub, String(i)), 10)).toBe(false);
    }
  });

  it('uses a random start by default → repeated solves differ but all verify', () => {
    const a = solvePoW(pub, 8);
    const b = solvePoW(pub, 8);
    expect(verifyPoW(pub, a, 8)).toBe(true);
    expect(verifyPoW(pub, b, 8)).toBe(true);
    // Astronomically unlikely to collide across the 2^30 start range.
    expect(a).not.toBe(b);
  });

  it('rejects a wrong nonce and the empty nonce', () => {
    expect(verifyPoW(pub, '0', 20)).toBe(false);
    expect(verifyPoW(pub, '', 8)).toBe(false);
  });
});

File: gateway/internal/api/pow_test.go
package api

import (
	"crypto/sha256"
	"strconv"
	"testing"
)

// TestHasLeadingZeroBits checks the bit-boundary logic byte- and bit-exactly,
// mirroring the contract in docs/10-api-contract.md (full = bits/8 zero bytes,
// then the next byte AND (0xff << (8-rem)) must be zero).
func TestHasLeadingZeroBits(t *testing.T) {
	cases := []struct {
		name   string
		digest []byte
		bits   int
		want   bool
	}{
		{"zero bits always true", []byte{0xff}, 0, true},
		{"one zero byte, 8 bits", []byte{0x00, 0xff}, 8, true},
		{"nonzero first byte, 8 bits", []byte{0x01, 0x00}, 8, false},
		{"4 leading zero bits ok", []byte{0x0f}, 4, true},
		{"3 leading zero bits, need 4", []byte{0x10}, 4, false},
		{"20 bits: two zero bytes + top nibble", []byte{0x00, 0x00, 0x0f}, 20, true},
		{"20 bits: third byte too big", []byte{0x00, 0x00, 0x10}, 20, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := hasLeadingZeroBits(tc.digest, tc.bits); got != tc.want {
				t.Errorf("hasLeadingZeroBits(%x, %d) = %v, want %v", tc.digest, tc.bits, got, tc.want)
			}
		})
	}
}

// TestSolvePoWRoundtrip solves the PoW for several pubkeys at a few
// difficulties and confirms the returned nonce is (a) a decimal string and
// (b) actually satisfies sha256(utf8(pubkey)||utf8(nonce)) with that many
// leading zero bits — the exact bytes every client language must hash.
func TestSolvePoWRoundtrip(t *testing.T) {
	pubkeys := []string{
		"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
		"AQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQE=",
		"3Qz8m2Yp1kL0aB9cD7eF6gH5iJ4kK3lM2nN1oO0pP8=",
	}
	// Keep difficulties low so the test is fast but still exercises the
	// byte and sub-byte boundaries (8 = one byte, 12 = byte+nibble).
	for _, bits := range []int{8, 12} {
		for _, pk := range pubkeys {
			nonce := solvePoW(pk, bits)

			if _, err := strconv.ParseUint(nonce, 10, 64); err != nil {
				t.Fatalf("nonce %q is not a decimal string: %v", nonce, err)
			}

			h := sha256.New()
			h.Write([]byte(pk))
			h.Write([]byte(nonce))
			if !hasLeadingZeroBits(h.Sum(nil), bits) {
				t.Fatalf("solved nonce %q does not satisfy %d bits for %q", nonce, bits, pk)
			}
		}
	}
}

// TestCheckPoWReplay verifies the server accepts a valid solution once and
// rejects a replay, an empty nonce, and a wrong nonce.
func TestCheckPoWReplay(t *testing.T) {
	s := &Server{powSeen: make(map[string]struct{})}
	const pk = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="

	if s.checkPoW(pk, "") {
		t.Error("empty nonce should be rejected")
	}
	if s.checkPoW(pk, "1") {
		t.Error("a nonce that does not meet the difficulty should be rejected")
	}

	nonce := solvePoW(pk, powBits)
	if !s.checkPoW(pk, nonce) {
		t.Fatalf("valid nonce %q rejected", nonce)
	}
	if s.checkPoW(pk, nonce) {
		t.Error("replayed nonce should be rejected the second time")
	}
}
